File: Examples/Chart/WinForm/CSharp/ChartGallery/TreeMap/NTreeMapUC.cs
using System;
using System.ComponentModel;
using System.Diagnostics;
using System.Drawing;
using System.Windows.Forms;
using Nevron.Chart;
using System.Xml;
using System.Reflection;
using System.IO;
using Nevron.UI;
using Nevron.GraphicsCore;
using Nevron.Chart.Windows;

namespace Nevron.Examples.Chart.WinForm
{
	[ToolboxItem(false)]
	public partial class NTreeMapUC : NExampleBaseUC
	{
		/// <summary> 
		/// Required designer variable.
		/// </summary>
		private System.ComponentModel.IContainer components = null;
		private UI.WinForm.Controls.NComboBox VerticalFillModeComboBox;
		private UI.WinForm.Controls.NComboBox HorizontalFillModeComboBox;
		private Label label1;
		private Label label2;
		private Label label3;
		private UI.WinForm.Controls.NComboBox ColorFillModeComboBox;
		private NTreeMapChart m_TreeMap;

		public NTreeMapUC()
		{
			InitializeComponent();
		}

		/// <summary> 
		/// Clean up any resources being used.
		/// </summary>
		/// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
		protected override void Dispose(bool disposing)
		{
			if (disposing && (components != null))
			{
				components.Dispose();
			}
			base.Dispose(disposing);
		}

		#region Component Designer generated code

		/// <summary> 
		/// Required method for Designer support - do not modify 
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
			this.VerticalFillModeComboBox = new Nevron.UI.WinForm.Controls.NComboBox();
			this.HorizontalFillModeComboBox = new Nevron.UI.WinForm.Controls.NComboBox();
			this.label1 = new System.Windows.Forms.Label();
			this.label2 = new System.Windows.Forms.Label();
			this.label3 = new System.Windows.Forms.Label();
			this.ColorFillModeComboBox = new Nevron.UI.WinForm.Controls.NComboBox();
			this.SuspendLayout();
			// 
			// VerticalFillModeComboBox
			// 
			this.VerticalFillModeComboBox.ListProperties.CheckBoxDataMember = "";
			this.VerticalFillModeComboBox.ListProperties.DataSource = null;
			this.VerticalFillModeComboBox.ListProperties.DisplayMember = "";
			this.VerticalFillModeComboBox.Location = new System.Drawing.Point(12, 34);
			this.VerticalFillModeComboBox.Name = "VerticalFillModeComboBox";
			this.VerticalFillModeComboBox.Size = new System.Drawing.Size(151, 21);
			this.VerticalFillModeComboBox.TabIndex = 6;
			this.VerticalFillModeComboBox.SelectedIndexChanged += new System.EventHandler(this.VerticalFillModeCombo_SelectedIndexChanged);
			// 
			// HorizontalFillModeComboBox
			// 
			this.HorizontalFillModeComboBox.ListProperties.CheckBoxDataMember = "";
			this.HorizontalFillModeComboBox.ListProperties.DataSource = null;
			this.HorizontalFillModeComboBox.ListProperties.DisplayMember = "";
			this.HorizontalFillModeComboBox.Location = new System.Drawing.Point(12, 85);
			this.HorizontalFillModeComboBox.Name = "HorizontalFillModeComboBox";
			this.HorizontalFillModeComboBox.Size = new System.Drawing.Size(151, 21);
			this.HorizontalFillModeComboBox.TabIndex = 7;
			this.HorizontalFillModeComboBox.SelectedIndexChanged += new System.EventHandler(this.HorizontalFillModeComboBox_SelectedIndexChanged);
			// 
			// label1
			// 
			this.label1.AutoSize = true;
			this.label1.Location = new System.Drawing.Point(12, 15);
			this.label1.Name = "label1";
			this.label1.Size = new System.Drawing.Size(90, 13);
			this.label1.TabIndex = 8;
			this.label1.Text = "Vertical Fill Mode:";
			// 
			// label2
			// 
			this.label2.AutoSize = true;
			this.label2.Location = new System.Drawing.Point(12, 65);
			this.label2.Name = "label2";
			this.label2.Size = new System.Drawing.Size(102, 13);
			this.label2.TabIndex = 9;
			this.label2.Text = "Horizontal Fill Mode:";
			// 
			// label3
			// 
			this.label3.AutoSize = true;
			this.label3.Location = new System.Drawing.Point(12, 118);
			this.label3.Name = "label3";
			this.label3.Size = new System.Drawing.Size(79, 13);
			this.label3.TabIndex = 11;
			this.label3.Text = "Color Fill Mode:";
			// 
			// ColorFillModeComboBox
			// 
			this.ColorFillModeComboBox.ListProperties.CheckBoxDataMember = "";
			this.ColorFillModeComboBox.ListProperties.DataSource = null;
			this.ColorFillModeComboBox.ListProperties.DisplayMember = "";
			this.ColorFillModeComboBox.Location = new System.Drawing.Point(12, 138);
			this.ColorFillModeComboBox.Name = "ColorFillModeComboBox";
			this.ColorFillModeComboBox.Size = new System.Drawing.Size(151, 21);
			this.ColorFillModeComboBox.TabIndex = 10;
			this.ColorFillModeComboBox.SelectedIndexChanged += new System.EventHandler(this.ColorFillModeComboBox_SelectedIndexChanged);
			// 
			// NTreeMapUC
			// 
			this.AutoScaleDimensions = new System.Drawing.SizeF(6F, 13F);
			this.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font;
			this.Controls.Add(this.label3);
			this.Controls.Add(this.ColorFillModeComboBox);
			this.Controls.Add(this.label2);
			this.Controls.Add(this.label1);
			this.Controls.Add(this.HorizontalFillModeComboBox);
			this.Controls.Add(this.VerticalFillModeComboBox);
			this.Name = "NTreeMapUC";
			this.Size = new System.Drawing.Size(178, 350);
			this.ResumeLayout(false);
			this.PerformLayout();

		}

		#endregion

		public override void Initialize()
		{
			base.Initialize();

			nChartControl1.Panels.Clear();
			nChartControl1.Controller.Tools.Add(new NTooltipTool());

			// set a chart title
			NLabel title = new NLabel("Tree Map");
			title.TextStyle.FontStyle = new NFontStyle("Times New Roman", 18, FontStyle.Italic);
			title.TextStyle.FillStyle = new NColorFillStyle(GreyBlue);
			title.DockMode = PanelDockMode.Top;
			nChartControl1.Panels.Add(title);
			
			m_TreeMap = new NTreeMapChart();
			nChartControl1.Panels.Add(m_TreeMap);
			m_TreeMap.DockMode = PanelDockMode.Fill;

			XmlDocument document = LoadData();

			foreach (XmlElement industry in document.DocumentElement)
			{
				NGroupTreeMapNode treeMapSeries = new NGroupTreeMapNode();

				treeMapSeries.StrokeStyle = new NStrokeStyle(4, Color.Black);
				treeMapSeries.Padding = new NMarginsL(2.0f);

				m_TreeMap.RootTreeMapNode.ChildNodes.Add(treeMapSeries);

				treeMapSeries.Label = industry.Attributes["Name"].Value;
				treeMapSeries.InteractivityStyle = new NInteractivityStyle(treeMapSeries.Label);

				foreach (XmlElement company in industry.ChildNodes)
				{
					double value = double.Parse(company.Attributes["Size"].Value);
					double change = double.Parse(company.Attributes["Change"].Value);
					string label = company.Attributes["Name"].Value;

					NValueTreeMapNode node = new NValueTreeMapNode(value, change, label);
					node.InteractivityStyle = new NInteractivityStyle(label);
					//						node.FillStyle = new NColorFillStyle(Color.Green);
					treeMapSeries.ChildNodes.Add(node);
				}
			}

			VerticalFillModeComboBox.FillFromEnum(typeof(TreeMapVerticalFillMode));
			VerticalFillModeComboBox.SelectedIndex = (int)TreeMapVerticalFillMode.Default;

			HorizontalFillModeComboBox.FillFromEnum(typeof(TreeMapHorizontalFillMode));
			HorizontalFillModeComboBox.SelectedIndex = (int)TreeMapVerticalFillMode.Default;

			ColorFillModeComboBox.Items.Add("Custom");
			ColorFillModeComboBox.Items.Add("Common Palette");
			ColorFillModeComboBox.Items.Add("Group Palette");
			ColorFillModeComboBox.SelectedIndex = 1;
		}
		/// <summary>
		/// Loads the data for the tree map
		/// </summary>
		/// <returns></returns>
		private XmlDocument LoadData()
		{
			Stream stream = NResourceHelper.GetResourceStream(GetType().Assembly, "TreeMapData.xml", "Nevron.Examples.Chart.WinForm.Resources");

			if (stream == null)
				return null;

			XmlDocument document = new XmlDocument();
			document.Load(stream);

			return document;
		}
	
		private void VerticalFillModeCombo_SelectedIndexChanged(object sender, EventArgs e)
		{
			m_TreeMap.RootTreeMapNode.VerticalFillMode = (TreeMapVerticalFillMode)VerticalFillModeComboBox.SelectedIndex;
			nChartControl1.Refresh();
		}

		private void HorizontalFillModeComboBox_SelectedIndexChanged(object sender, EventArgs e)
		{
			m_TreeMap.RootTreeMapNode.HorizontalFillMode = (TreeMapHorizontalFillMode)HorizontalFillModeComboBox.SelectedIndex;
			nChartControl1.Refresh();
		}

		private void ColorFillModeComboBox_SelectedIndexChanged(object sender, EventArgs e)
		{
			switch (ColorFillModeComboBox.SelectedIndex)
			{
				case 0:
					{
						// custom color filling -> assign colors to each group (child nodes will inherit that fill)
						NChartPalette palette = new NChartPalette(ChartPredefinedPalette.Bright);

						int industryIndex = 0;
						foreach (NGroupTreeMapNode industryTreeMapNode in m_TreeMap.RootTreeMapNode.ChildNodes)
						{
							foreach (NTreeMapNode companyTreeMapNode in industryTreeMapNode.ChildNodes)
							{
								companyTreeMapNode.FillStyle = new NColorFillStyle(palette.SeriesColors[(industryIndex) % palette.SeriesColors.Count]);
							}

							industryIndex++;
						}
					}
					break;
				case 1:
					{
						// palette filling -> remove all fill styles assigned to nodes
						foreach (NGroupTreeMapNode industryTreeMapNode in m_TreeMap.RootTreeMapNode.ChildNodes)
						{
							industryTreeMapNode.Palette = null;

							foreach (NTreeMapNode companyTreeMapNode in industryTreeMapNode.ChildNodes)
							{
								companyTreeMapNode.FillStyle = null;
							}
						}
					}
					break;
				case 2:
					{
						// palette filling -> remove all fill styles assigned to nodes
						foreach (NGroupTreeMapNode industryTreeMapNode in m_TreeMap.RootTreeMapNode.ChildNodes)
						{
							NPalette palette = new NPalette();
							palette.Mode = PaletteMode.AutoMinMaxColor;
							industryTreeMapNode.Palette = palette;

							foreach (NTreeMapNode companyTreeMapNode in industryTreeMapNode.ChildNodes)
							{
								companyTreeMapNode.FillStyle = null;
							}
						}
					}
					break;
			}

			nChartControl1.Refresh();
		}
	}
}

File: Examples/Chart/WinForm/VB.NET/ChartGallery/TreeMap/NTreeMapUC.vb
Imports System
Imports System.ComponentModel
Imports System.Diagnostics
Imports System.Drawing
Imports System.Windows.Forms
Imports Nevron.Chart
Imports System.Xml
Imports System.Reflection
Imports System.IO
Imports Nevron.UI
Imports Nevron.GraphicsCore
Imports Nevron.Chart.Windows

Namespace Nevron.Examples.Chart.WinForm
	<ToolboxItem(False)>
	Partial Public Class NTreeMapUC
		Inherits NExampleBaseUC

		''' <summary> 
		''' Required designer variable.
		''' </summary>
		Private components As System.ComponentModel.IContainer = Nothing
		Private WithEvents VerticalFillModeComboBox As UI.WinForm.Controls.NComboBox
		Private WithEvents HorizontalFillModeComboBox As UI.WinForm.Controls.NComboBox
		Private label1 As Label
		Private label2 As Label
		Private label3 As Label
		Private WithEvents ColorFillModeComboBox As UI.WinForm.Controls.NComboBox
		Private m_TreeMap As NTreeMapChart

		Public Sub New()
			InitializeComponent()
		End Sub

		''' <summary> 
		''' Clean up any resources being used.
		''' </summary>
		''' <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
		Protected Overrides Sub Dispose(ByVal disposing As Boolean)
			If disposing AndAlso (components IsNot Nothing) Then
				components.Dispose()
			End If
			MyBase.Dispose(disposing)
		End Sub

		#Region "Component Designer generated code"

		''' <summary> 
		''' Required method for Designer support - do not modify 
		''' the contents of this method with the code editor.
		''' </summary>
		Private Sub InitializeComponent()
			Me.VerticalFillModeComboBox = New Nevron.UI.WinForm.Controls.NComboBox()
			Me.HorizontalFillModeComboBox = New Nevron.UI.WinForm.Controls.NComboBox()
			Me.label1 = New System.Windows.Forms.Label()
			Me.label2 = New System.Windows.Forms.Label()
			Me.label3 = New System.Windows.Forms.Label()
			Me.ColorFillModeComboBox = New Nevron.UI.WinForm.Controls.NComboBox()
			Me.SuspendLayout()
			' 
			' VerticalFillModeComboBox
			' 
			Me.VerticalFillModeComboBox.ListProperties.CheckBoxDataMember = ""
			Me.VerticalFillModeComboBox.ListProperties.DataSource = Nothing
			Me.VerticalFillModeComboBox.ListProperties.DisplayMember = ""
			Me.VerticalFillModeComboBox.Location = New System.Drawing.Point(12, 34)
			Me.VerticalFillModeComboBox.Name = "VerticalFillModeComboBox"
			Me.VerticalFillModeComboBox.Size = New System.Drawing.Size(151, 21)
			Me.VerticalFillModeComboBox.TabIndex = 6
'INSTANT VB NOTE: The following InitializeComponent event wireup was converted to a 'Handles' clause:
'ORIGINAL LINE: this.VerticalFillModeComboBox.SelectedIndexChanged += new System.EventHandler(this.VerticalFillModeCombo_SelectedIndexChanged);
			' 
			' HorizontalFillModeComboBox
			' 
			Me.HorizontalFillModeComboBox.ListProperties.CheckBoxDataMember = ""
			Me.HorizontalFillModeComboBox.ListProperties.DataSource = Nothing
			Me.HorizontalFillModeComboBox.ListProperties.DisplayMember = ""
			Me.HorizontalFillModeComboBox.Location = New System.Drawing.Point(12, 85)
			Me.HorizontalFillModeComboBox.Name = "HorizontalFillModeComboBox"
			Me.HorizontalFillModeComboBox.Size = New System.Drawing.Size(151, 21)
			Me.HorizontalFillModeComboBox.TabIndex = 7
'INSTANT VB NOTE: The following InitializeComponent event wireup was converted to a 'Handles' clause:
'ORIGINAL LINE: this.HorizontalFillModeComboBox.SelectedIndexChanged += new System.EventHandler(this.HorizontalFillModeComboBox_SelectedIndexChanged);
			' 
			' label1
			' 
			Me.label1.AutoSize = True
			Me.label1.Location = New System.Drawing.Point(12, 15)
			Me.label1.Name = "label1"
			Me.label1.Size = New System.Drawing.Size(90, 13)
			Me.label1.TabIndex = 8
			Me.label1.Text = "Vertical Fill Mode:"
			' 
			' label2
			' 
			Me.label2.AutoSize = True
			Me.label2.Location = New System.Drawing.Point(12, 65)
			Me.label2.Name = "label2"
			Me.label2.Size = New System.Drawing.Size(102, 13)
			Me.label2.TabIndex = 9
			Me.label2.Text = "Horizontal Fill Mode:"
			' 
			' label3
			' 
			Me.label3.AutoSize = True
			Me.label3.Location = New System.Drawing.Point(12, 118)
			Me.label3.Name = "label3"
			Me.label3.Size = New System.Drawing.Size(79, 13)
			Me.label3.TabIndex = 11
			Me.label3.Text = "Color Fill Mode:"
			' 
			' ColorFillModeComboBox
			' 
			Me.ColorFillModeComboBox.ListProperties.CheckBoxDataMember = ""
			Me.ColorFillModeComboBox.ListProperties.DataSource = Nothing
			Me.ColorFillModeComboBox.ListProperties.DisplayMember = ""
			Me.ColorFillModeComboBox.Location = New System.Drawing.Point(12, 138)
			Me.ColorFillModeComboBox.Name = "ColorFillModeComboBox"
			Me.ColorFillModeComboBox.Size = New System.Drawing.Size(151, 21)
			Me.ColorFillModeComboBox.TabIndex = 10
'INSTANT VB NOTE: The following InitializeComponent event wireup was converted to a 'Handles' clause:
'ORIGINAL LINE: this.ColorFillModeComboBox.SelectedIndexChanged += new System.EventHandler(this.ColorFillModeComboBox_SelectedIndexChanged);
			' 
			' NTreeMapUC
			' 
			Me.AutoScaleDimensions = New System.Drawing.SizeF(6F, 13F)
			Me.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font
			Me.Controls.Add(Me.label3)
			Me.Controls.Add(Me.ColorFillModeComboBox)
			Me.Controls.Add(Me.label2)
			Me.Controls.Add(Me.label1)
			Me.Controls.Add(Me.HorizontalFillModeComboBox)
			Me.Controls.Add(Me.VerticalFillModeComboBox)
			Me.Name = "NTreeMapUC"
			Me.Size = New System.Drawing.Size(178, 350)
			Me.ResumeLayout(False)
			Me.PerformLayout()

		End Sub

		#End Region

		Public Overrides Sub Initialize()
			MyBase.Initialize()

			nChartControl1.Panels.Clear()
			nChartControl1.Controller.Tools.Add(New NTooltipTool())

			' set a chart title
			Dim title As New NLabel("Tree Map")
			title.TextStyle.FontStyle = New NFontStyle("Times New Roman", 18, FontStyle.Italic)
			title.TextStyle.FillStyle = New NColorFillStyle(GreyBlue)
			title.DockMode = PanelDockMode.Top
			nChartControl1.Panels.Add(title)

			m_TreeMap = New NTreeMapChart()
			nChartControl1.Panels.Add(m_TreeMap)
			m_TreeMap.DockMode = PanelDockMode.Fill

			Dim document As XmlDocument = LoadData()

			For Each industry As XmlElement In document.DocumentElement
				Dim treeMapSeries As New NGroupTreeMapNode()

				treeMapSeries.StrokeStyle = New NStrokeStyle(4, Color.Black)
				treeMapSeries.Padding = New NMarginsL(2.0F)

				m_TreeMap.RootTreeMapNode.ChildNodes.Add(treeMapSeries)

				treeMapSeries.Label = industry.Attributes("Name").Value
				treeMapSeries.InteractivityStyle = New NInteractivityStyle(treeMapSeries.Label)

				For Each company As XmlElement In industry.ChildNodes
					Dim value As Double = Double.Parse(company.Attributes("Size").Value)
					Dim change As Double = Double.Parse(company.Attributes("Change").Value)
					Dim label As String = company.Attributes("Name").Value

					Dim node As New NValueTreeMapNode(value, change, label)
					node.InteractivityStyle = New NInteractivityStyle(label)
					'						node.FillStyle = new NColorFillStyle(Color.Green);
					treeMapSeries.ChildNodes.Add(node)
				Next company
			Next industry

			VerticalFillModeComboBox.FillFromEnum(GetType(TreeMapVerticalFillMode))
			VerticalFillModeComboBox.SelectedIndex = CInt(TreeMapVerticalFillMode.Default)

			HorizontalFillModeComboBox.FillFromEnum(GetType(TreeMapHorizontalFillMode))
			HorizontalFillModeComboBox.SelectedIndex = CInt(TreeMapVerticalFillMode.Default)

			ColorFillModeComboBox.Items.Add("Custom")
			ColorFillModeComboBox.Items.Add("Common Palette")
			ColorFillModeComboBox.Items.Add("Group Palette")
			ColorFillModeComboBox.SelectedIndex = 1
		End Sub
		''' <summary>
		''' Loads the data for the tree map
		''' </summary>
		''' <returns></returns>
		Private Function LoadData() As XmlDocument
			Dim stream As Stream = NResourceHelper.GetResourceStream(Me.GetType().Assembly, "TreeMapData.xml", "Nevron.Examples.Chart.WinForm.Resources")

			If stream Is Nothing Then
				Return Nothing
			End If

			Dim document As New XmlDocument()
			document.Load(stream)

			Return document
		End Function

		Private Sub VerticalFillModeCombo_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles VerticalFillModeComboBox.SelectedIndexChanged
			m_TreeMap.RootTreeMapNode.VerticalFillMode = CType(VerticalFillModeComboBox.SelectedIndex, TreeMapVerticalFillMode)
			nChartControl1.Refresh()
		End Sub

		Private Sub HorizontalFillModeComboBox_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles HorizontalFillModeComboBox.SelectedIndexChanged
			m_TreeMap.RootTreeMapNode.HorizontalFillMode = CType(HorizontalFillModeComboBox.SelectedIndex, TreeMapHorizontalFillMode)
			nChartControl1.Refresh()
		End Sub

		Private Sub ColorFillModeComboBox_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles ColorFillModeComboBox.SelectedIndexChanged
			Select Case ColorFillModeComboBox.SelectedIndex
				Case 0
						' custom color filling -> assign colors to each group (child nodes will inherit that fill)
						Dim palette As New NChartPalette(ChartPredefinedPalette.Bright)

						Dim industryIndex As Integer = 0
						For Each industryTreeMapNode As NGroupTreeMapNode In m_TreeMap.RootTreeMapNode.ChildNodes
							For Each companyTreeMapNode As NTreeMapNode In industryTreeMapNode.ChildNodes
								companyTreeMapNode.FillStyle = New NColorFillStyle(palette.SeriesColors((industryIndex) Mod palette.SeriesColors.Count))
							Next companyTreeMapNode

							industryIndex += 1
						Next industryTreeMapNode
				Case 1
						' palette filling -> remove all fill styles assigned to nodes
						For Each industryTreeMapNode As NGroupTreeMapNode In m_TreeMap.RootTreeMapNode.ChildNodes
							industryTreeMapNode.Palette = Nothing

							For Each companyTreeMapNode As NTreeMapNode In industryTreeMapNode.ChildNodes
								companyTreeMapNode.FillStyle = Nothing
							Next companyTreeMapNode
						Next industryTreeMapNode
				Case 2
						' palette filling -> remove all fill styles assigned to nodes
						For Each industryTreeMapNode As NGroupTreeMapNode In m_TreeMap.RootTreeMapNode.ChildNodes
							Dim palette As New NPalette()
							palette.Mode = PaletteMode.AutoMinMaxColor
							industryTreeMapNode.Palette = palette

							For Each companyTreeMapNode As NTreeMapNode In industryTreeMapNode.ChildNodes
								companyTreeMapNode.FillStyle = Nothing
							Next companyTreeMapNode
						Next industryTreeMapNode
			End Select

			nChartControl1.Refresh()
		End Sub
	End Class
End Namespace
